File: julia/benchmark.jl
include(joinpath(@__DIR__, "mvnormal.jl"))

const COMPARISON_NORMAL_SEED = 0x5EED2021

function parse_args(args)
    values = Dict{String, Int}(
        "dim" => 10,
        "samples" => 10_000,
        "repeats" => 5,
    )
    normal = :julia
    batch = false
    i = 1
    while i <= length(args)
        arg = args[i]
        startswith(arg, "--") || throw(ArgumentError("unexpected argument: $arg"))

        if arg == "--batch"
            batch = true
            i += 1
            continue
        end

        key, value = if occursin('=', arg)
            split(arg[3:end], "=", limit=2) |> Tuple
        else
            i < length(args) || throw(ArgumentError("missing value for $arg"))
            (arg[3:end], args[i + 1])
        end
        if key == "normal"
            value in ("julia", "polar", "ziggurat") ||
                throw(ArgumentError("--normal must be julia, polar, or ziggurat"))
            normal = Symbol(value)
            i += occursin('=', arg) ? 1 : 2
            continue
        end
        haskey(values, key) || throw(ArgumentError("unknown option: --$key"))
        parsed = try
            parse(Int, value)
        catch
            throw(ArgumentError("--$key must be an integer"))
        end
        parsed >= 0 || throw(ArgumentError("--$key must be non-negative"))
        values[key] = parsed
        i += occursin('=', arg) ? 1 : 2
    end

    values["dim"] > 0 || throw(ArgumentError("--dim must be positive"))
    values["samples"] > 0 || throw(ArgumentError("--samples must be positive"))
    values["repeats"] > 0 || throw(ArgumentError("--repeats must be positive"))
    return (values=values, normal=normal, batch=batch)
end

function main(args)
    # Keep the comparison single-threaded even though Julia's `mul!` would
    # otherwise use a multithreaded BLAS for the batched path.
    BLAS.set_num_threads(1)

    parsed = parse_args(args)
    options = parsed.values
    normal = parsed.normal
    batch = parsed.batch
    dim = options["dim"]
    nsamples = options["samples"]
    repeats = options["repeats"]

    normal === :julia && Random.seed!(0xc0ffee)
    μ = [0.01 * (index - 1) for index in 1:dim]
    Σ = [ldexp(1.0, -2 * abs(row - column)) for row in 1:dim, column in 1:dim]

    setup_start = time_ns()
    d = MvNormal(μ, Σ)
    setup_sec = (time_ns() - setup_start) / 1.0e9

    rng = if normal === :julia
        Random.default_rng()
    elseif normal === :polar
        MarsagliaPolarRNG(COMPARISON_NORMAL_SEED)
    else
        ZigguratRNG(COMPARISON_NORMAL_SEED)
    end
    warmup_rng = if normal === :julia
        Random.default_rng()
    elseif normal === :polar
        MarsagliaPolarRNG(0xabad1dea)
    else
        ZigguratRNG(0xabad1dea)
    end

    sample_times = Vector{Float64}(undef, repeats)
    checksum = 0.0

    if batch
        out = Matrix{Float64}(undef, dim, nsamples)

        # Compile the batched sampling path before timing it.
        sample!(warmup_rng, d, out)
        warmup_checksum = sum(out)

        for repeat in 1:repeats
            start = time_ns()
            sample!(rng, d, out)
            sample_times[repeat] = (time_ns() - start) / 1.0e9
            checksum += sum(out)
        end
    else
        # Compile the per-sample sampling path before timing it.
        out = Vector{Float64}(undef, dim)
        warmup_checksum = 0.0
        for _ in 1:nsamples
            sample!(warmup_rng, d, out)
            warmup_checksum += sum(out)
        end

        for repeat in 1:repeats
            start = time_ns()
            for _ in 1:nsamples
                sample!(rng, d, out)
                checksum += sum(out)
            end
            sample_times[repeat] = (time_ns() - start) / 1.0e9
        end
    end

    # Keep the warmup result observable without including it in the reported checksum.
    checksum += 0.0 * warmup_checksum
    avg_sample_sec = sum(sample_times) / repeats
    min_sample_sec = minimum(sample_times)
    label = (normal === :julia ? "julia" : "julia-$normal") * (batch ? "-batch" : "")
    println("$label,$dim,$nsamples,$repeats,$setup_sec,$avg_sample_sec,$min_sample_sec,$checksum")
end

main(ARGS)

File: python/benchmark.py
#!/usr/bin/env python3
"""Benchmark the Python MvNormal implementation under the common conditions.

Run with the project environment, for example:

    uv run --project python python/benchmark.py --dim 64 --samples 20000 --repeats 5
"""

from __future__ import annotations

import argparse
import time

import numpy as np

from mvnormal import MvNormal, NormalAlgorithm, NormalRng

COMPARISON_NORMAL_SEED = 0x5EED2021
WARMUP_SEED = 0xABAD1DEA


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="MvNormal Python benchmark")
    parser.add_argument("--dim", type=int, default=10)
    parser.add_argument("--samples", type=int, default=10_000)
    parser.add_argument("--repeats", type=int, default=5)
    parser.add_argument(
        "--normal",
        choices=("polar", "ziggurat"),
        default="polar",
        help="standard normal algorithm (default: polar)",
    )
    arguments = parser.parse_args()
    if arguments.dim <= 0:
        parser.error("--dim must be positive")
    if arguments.samples <= 0:
        parser.error("--samples must be positive")
    if arguments.repeats <= 0:
        parser.error("--repeats must be positive")
    return arguments


def build_inputs(dimension: int) -> tuple[np.ndarray, np.ndarray]:
    indices = np.arange(dimension, dtype=np.float64)
    mean = 0.01 * indices
    covariance = 0.25 ** np.abs(indices[:, None] - indices[None, :])
    return mean, covariance


def main() -> None:
    arguments = parse_args()
    algorithm = NormalAlgorithm.MARSAGLIA_POLAR
    if arguments.normal == "ziggurat":
        algorithm = NormalAlgorithm.ZIGGURAT

    mean, covariance = build_inputs(arguments.dim)

    setup_start = time.perf_counter()
    distribution = MvNormal(mean, covariance)
    setup_sec = time.perf_counter() - setup_start

    # Compile the Numba kernels before timing and keep the comparison seed
    # untouched by using a separate generator for the warmup.
    warmup_rng = NormalRng(WARMUP_SEED, algorithm)
    distribution.sample_checksum(warmup_rng, min(arguments.samples, 64))

    rng = NormalRng(COMPARISON_NORMAL_SEED, algorithm)
    durations = []
    checksum = 0.0
    for _ in range(arguments.repeats):
        start = time.perf_counter()
        checksum += distribution.sample_checksum(rng, arguments.samples)
        durations.append(time.perf_counter() - start)

    avg_sample_sec = sum(durations) / arguments.repeats
    min_sample_sec = min(durations)
    print(
        f"python-{algorithm.label()},{arguments.dim},{arguments.samples},"
        f"{arguments.repeats},{setup_sec:.9g},{avg_sample_sec:.9g},"
        f"{min_sample_sec:.9g},{checksum!r}"
    )


if __name__ == "__main__":
    main()
